File: estructurales/proxy/ProxyExample.java
interface Image {
    void display();
}

class RealImage implements Image {
    private final String filename;

    public RealImage(String filename) {
        this.filename = filename;
        loadFromDisk();
    }

    private void loadFromDisk() {
        System.out.println("Loading " + filename + " from disk...");
    }

    @Override
    public void display() {
        System.out.println("Displaying " + filename);
    }
}

class ImageProxy implements Image {
    private final String filename;
    private RealImage real; // se crea solo cuando se usa

    public ImageProxy(String filename) {
        this.filename = filename;
    }

    @Override
    public void display() {
        if (real == null) {
            real = new RealImage(filename); // lazy load
        }
        real.display();
    }
}

public class ProxyExample {
    public static void main(String[] args) {
        Image img = new ImageProxy("photo.png");
        img.display();
        img.display();
    }
}

File: estructurales/proxy/ProxyExample.py
class Image:
    def display(self) -> None:
        raise NotImplementedError


class RealImage(Image):
    def __init__(self, filename: str) -> None:
        self.filename = filename
        self._load_from_disk()

    def _load_from_disk(self) -> None:
        print(f"Loading {self.filename} from disk...")

    def display(self) -> None:
        print(f"Displaying {self.filename}")


class ImageProxy(Image):
    def __init__(self, filename: str) -> None:
        self.filename = filename
        self._real: RealImage | None = None

    def display(self) -> None:
        # Lazy load: solo crea RealImage cuando se necesita
        if self._real is None:
            self._real = RealImage(self.filename)
        self._real.display()


if __name__ == "__main__":
    img = ImageProxy("photo.png")
    img.display()  # carga + muestra
    img.display()  # solo muestra (ya está cargada)
